rustler_codegen/src/ex_struct.rs

```diff
@@ -55,23 +55,20 @@ pub fn transcoder_decorator(ast: &syn::DeriveInput) -> TokenStream {
 fn gen_decoder(ctx: &Context, fields: &[&Field], atoms_module_name: &Ident) -> TokenStream {
     let struct_type = &ctx.ident_with_lifetime;
     let struct_name = ctx.ident;
+    let struct_name_str = struct_name.to_string();
 
-    let field_defs: Vec<TokenStream> = fields
+    let idents: Vec<_> = fields
         .iter()
-        .map(|field| {
-            let ident = field.ident.as_ref().unwrap();
+        .map(|field| field.ident.as_ref().unwrap())
+        .collect();
+
+    let field_defs: Vec<TokenStream> = idents
+        .iter()
+        .map(|ident| {
             let ident_str = ident.to_string();
             let atom_fun = Ident::new(&format!("atom_{}", ident_str), Span::call_site());
-            let error_message = format!(
-                "Could not decode field :{} on %{}{{}}",
-                ident.to_string(),
-                struct_name.to_string()
-            );
             quote! {
-                #ident: match ::rustler::Decoder::decode(term.map_get(#atom_fun().encode(env))?) {
-                    Err(_) => return Err(::rustler::Error::RaiseTerm(Box::new(#error_message))),
-                    Ok(value) => value
-                }
+                let #ident = try_decode_field(env, term, #atom_fun())?;
```

Comment:
Just so things are a bit more robust, maybe we should be using a number index or something as the variable name in the generated code?

Reply:
01d749c. Note that I am unsure if we actually need to do this, as we have the complete function under our control here (we generate it right now). In any case, it shouldn't hurt.

Reply:
This conflicts with #303. When a field is named `r#override`, the generated name is something like `RUSTLER_struct_field_0_r#override`, which isn't a valid identifier.

Reply:
I extracted generating the identifiers to the `Context` module in 95520dd.